Multiple enhancements

icereed opened the pull request:
This would solve #4 and #3.
Additionally you may skip folder for exporting.

### Summary of new features

1. Able to provide the flags via ENV variable
2. Able to specify https URLs
3. Add credential management (decrypt and apply)
4. Added flag to ignore folders
5. Added test cases
6. Enable import/export into folders

### Summary of fixes

1. Fix not to stop importing if one job fails

icereed committed:
Able to specify https URLs

icereed committed:
Added test case for getSanitizedUrl()

icereed committed:
Ability to see if exported Job is folder

icereed committed:
Added flag to ignore folders

icereed committed:
Able to provide the flags via ENV variable

ambakshi commented:
Oh man, I just went to solve the same problem :P

ambakshi commented on main.go:
```diff
@@ -181,3 +181,11 @@ func main() {
+
+func GetSanitizedUrl(url string) string {
+	if !strings.HasPrefix(url, "https://") && !strings.HasPrefix(url, "http://") {
+		url = "http://" + url
```
This has an issue in that passing an empty string will return "http://". This causes the subsequent check for if server == "" to go through.

icereed replied:
Thanks! Good catch. Fixed that one and added a fitting test case for the GetSanitizedUrl() function.

icereed committed:
Fixed a minor bug where the CLI would ignore empty server URL

icereed commented:
@mlabouardy this PR looks ready. Would you like to draft the new version or should I do it in this PR?

icereed committed:
Enable import/export into folders

Previously it was not possible to export and import from and to folders.
In order to achieve this a new parameter `--folder` is introduced.
Given this parameter it will resolve the URL of the folder and
export/import jobs there.

In order to achieve this, the API endpoint /api/xml of a folder is used
(also the for the "root" folder) instead of /view/all/api/xml in order
not to rely on special view, which is only existing in root.

Also a bug is fixed if an old version of Jenkins does not
support crumb issueing.

icereed commented:
@mlabouardy Could you take a look please?

icereed committed:
Add credential management (decrypt and apply)

This enables migration of credentials between Jenkins instances.

Butler will get the XML config of any folder and will ingest
the encrypted secrets. The encrypted secrets are then decrypted
via a Groovy script executed on Jenkins master. This needs
most of the times admin privileges.

In a second step you can pipe in the decrypted credentials
and apply them to another folder or even instance.
Also a Groovy script is leveraged in order to achieve this.

icereed committed:
Merge pull request #1 from kentrikos/feature/credential-mover

This enables migration of credentials between Jenkins instances.

Butler will get the XML config of any folder and will ingest
the encrypted secrets. The encrypted secrets are then decrypted
via a Groovy script executed on Jenkins master. This needs
most of the times admin privileges.

In a second step you can pipe in the decrypted credentials
and apply them to another folder or even instance.
Also a Groovy script is leveraged in order to achieve this.

icereed committed:
Fix not to stop importing if one job fails

Butler stopped importing jobs if any job was existing
already or was malformatted.
Now the error is printed out, but the
import will keep working.

icereed committed:
fix: properly encode secret files for transfer

This was needed, because special characters we're not correctly interpreted.

r-divakaran-hrs commented:
Can we please merge this !?

icereed committed:
fix: Actually printing error if crumb issueing is not enabled

icereed committed:
Add possibility to (recursively) list folders

This feature is useful if you want to export a folder tree.

Also made the first step into refactoring the HTTP request handling.

icereed committed:
Add ability to specify server via ENV var

This is handy if you want to avoid typing the server name.

icereed commented:
is this repo not maintained?

mlabouardy commented:
@icereed @ambakshi sorry for late response guys

icereed commented:
The pipeline fails, because there are no working credentials for S3 in Circle CI.

mlabouardy commented:
@icereed fixed

icereed commented:
Feature is released :party:
